### PR TITLE
Csr jalr

### PR DESCRIPTION
**Pull request by deveshshevde**
Screenshot 2024-10-29 at 12 14 28 PM

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -63,7 +64,7 @@ debug:
 ASMDIR ?= rv32_asm
 asm:
 	mkdir -p ./$(BUILD_DIR)
-	riscv64-unknown-elf-gcc -march=rv32i -mabi=ilp32 -static -nostdlib -T$(ASMDIR)/$(LD_FILE) $(ASMDIR)/$(ASM).s -o $(BUILD_DIR)/${ASM}.elf
+	riscv64-unknown-elf-gcc -march=rv32i_zicsr -mabi=ilp32 -static -nostdlib -T$(ASMDIR)/$(LD_FILE) $(ASMDIR)/$(ASM).s -o $(BUILD_DIR)/${ASM}.elf
```
Comment (ghost) [inline Makefile:67]: zicsr is implicit. You can use rv32i

File: inc/elf_reader.h [hunk 2]
```diff
@@ -1,7 +1,6 @@
 #ifndef _ELF_READER_H_
 #define _ELF_READER_H_
-
-#include <elf.h>
+#include "elf.h"
```
Comment (ghost) [inline inc/elf_reader.h:3]: Revert this change, earlier one is correct.

File: rv32_asm/add.s [hunk 3]
```diff
@@ -8,10 +8,14 @@ _start:
...
     li t0, 0x13
     li t1, 0x23
+    csrr a0, mcause   
+    csrw mstatus, t6
+    csrr a1 ,misa 
     add t2, t0, t1
+    li       t1, 0x80               
     la t4, msg
-
+
```
Comment (ghost) [inline rv32_asm/add.s:17]: Whitespace...

File: src/opcode/rv32_fetch_decode.c [hunk 4]
```diff
@@ -166,20 +212,23 @@ void rv32_decode(uint32_t word, ram_t *ram)
...
+
+    args.csr_ctx =  csr_list;
     if(!exec_cb)
     {
         printf("\n[E]: No callback regiterd to perform execution\n");
         g_rv32i_ctx += RV32_PC_JUMP;
         return;
     }
-    g_rv32i_ctx->pc = exec_cb(&args);
+    g_rv32i_ctx->pc = exec_cb(&args); //
```
Comment (ghost) [inline src/opcode/rv32_fetch_decode.c:231]: ??

File: src/opcode/rv32_fetch_decode.c [hunk 5]
```diff
@@ -1,26 +1,66 @@
...
+
+
 static uint32_t get_i(uint32_t wc)
 {
     wc = RV_MASK_I(wc);
     return wc & (1 << 12) ? wc | 0xFFFFF000 : wc & 0x00000FFF;
 }
 
-static void decode_and_print(const char *template, uint32_t rd, uint32_t r1, uint32_t r2, uint32_t imm)
+static void decode_and_print(const char *template, uint32_t rd, uint32_t r1, uint32_t r2, uint32_t imm , uint32_t csr_index)
```
Comment (ghost) [inline src/opcode/rv32_fetch_decode.c:63]: space after `imm ,`

File: src/opcode/rv32_fetch_decode.c [hunk 6]
```diff
@@ -1,26 +1,66 @@
 #include "main.h"
-
+#include "rv32I_struct.h"
```
Comment (ghost) [inline src/opcode/rv32_fetch_decode.c:2]: it should be in `<...>`

File: Makefile (hunk 1 above)
Comment (deveshshevde) [reply Makefile:67]: ```rv32_asm/csr.s:6: Error: unrecognized opcode `csrr a3,mcause', extension `zicsr' required```
For mac builds it is showing error

**Comment by pranjalchanda08** [comment]
I Think this is stale PR. @deveshshevde, please consider doing the recommended changes and raise a new PR.

File: Makefile (hunk 1 above)
Comment (deveshshevde) [reply Makefile:67]: @pranjalchanda08  @akashkollipara  what about this?

Comment (ghost) [reply Makefile:67]: It looks like the newer toolchain needs this.
